### PR TITLE
[O2B-1274] Api for fetching GAQ flags with views

### PR DESCRIPTION
**Pull request by xsalonx**
#### I have a JIRA ticket
- [x] branch and/or PR name(s) include(s) JIRA ID
- [ ] issue has "Fix version" assigned
- [ ] issue "Status" is set to "In review"
- [x] PR labels are selected

Notable changes for users:
- NA

Notable changes for developers:
- Added methods for fetching QC flags in GAQ periods

Changes made to the database:
- NA

### CODE REVIEW

File: lib/server/controllers/qcFlag.controller.js [hunk 1]
```diff
@@ -248,6 +248,34 @@ const getQcFlagsSummaryHandler = async (req, res) => {
...
+        })),
+        req,
+        res,
+    );
+    if (validatedDTO) {
+        try {
+            const {
+                dataPassId,
+                runNumber,
+            } = validatedDTO.query;
```
Comment (martinboulais) [inline lib/server/controllers/qcFlag.controller.js:269]: No need for newline here

File: lib/server/services/qualityControlFlag/QcFlagService.js [hunk 2]
```diff
@@ -22,7 +22,11 @@ const {
     },
 } = require('../../../database/index.js');
 const { dataSource } = require('../../../database/DataSource.js');
-const { qcFlagAdapter, dataPassQcFlagAdapter, simulationPassQcFlagAdapter } = require('../../../database/adapters/index.js');
+const {
+    qcFlagAdapter,
+    dataPassQcFlagAdapter,
+    simulationPassQcFlagAdapter,
+} = require('../../../database/adapters/index.js');
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagService.js:29]: Why this change?

File: lib/server/services/qualityControlFlag/QcFlagService.js [hunk 3]
```diff
@@ -546,6 +550,32 @@ class QcFlagService {
...
 
+    /**
+     * Find QC flags in GAQ effective periods for given data pass and run
+     *
+     * @param {number} dataPassId id od data pass
+     * @param {number} runNumber run number
+     * @return {Promise<GaqQcFlags>} promise of aggregated QC flags
+     */
+    async getGaqQcFlags(dataPassId, runNumber) {
+        const effectiveGaqPeriods = await QcFlagRepository.findFlagsInGaqEffectivePeriods({ dataPassId, runNumbers: [runNumber] });
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagService.js:561]: why is the variable named `effectiveGaqPeriods` ?

File: lib/database/repositories/QcFlagRepository.js [hunk 4]
```diff
@@ -26,6 +36,93 @@ class QcFlagRepository extends Repository {
...
     }
 
+    /**
+     * Find IDs of QC flags in GAQ effective periods for given data pass and runs
+     *
+     * @param {number} scope.dataPassId id of data pass id
+     * @param {number[]} [scope.runNumbers] list of run numbers
+     * @return {Promise<GaqEffectivePeriods>} promise resolvesd once periods are calculated
+     */
+    async findFlagsInGaqEffectivePeriods({ dataPassId, runNumbers }) {
```
Comment (martinboulais) [inline lib/database/repositories/QcFlagRepository.js:46]: If this function is named `findFlagsInGqaEffectivePeriods` it should return QcFlags or SequelizeQcFlags?

File: lib/server/services/qualityControlFlag/QcFlagService.js [hunk 5]
```diff
@@ -546,6 +550,32 @@ class QcFlagService {
...
+        const qcFlags = (await QcFlagRepository.findAll({
+            where: { id: { [Op.in]: effectiveGaqPeriods.flatMap(({ contributingFlagIds }) => contributingFlagIds) } },
+        })).map(qcFlagAdapter.toEntity);
+
+        const idToFlag = Object.fromEntries(qcFlags.map((flag) => [flag.id, flag]));
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagService.js:566]: why doing this?
Why not extracting a list of flag ids, then simply fetching it and using this as `contributingFlags` ?

File: lib/database/adapters/dpl/DplDetectorAdapter.js [hunk 6]
```diff
@@ -59,6 +59,24 @@ class DplDetectorAdapter {
             processesExecutions: processesExecutions.map(this.dplProcessExecutionAdapter.toDatabase),
         };
     }
+
+    /**
+     * Converts the given database object to an minified entity object.
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/database/adapters/dpl/DplDetectorAdapter.js:64]: yes, fixed

File: lib/database/repositories/QcFlagRepository.js [hunk 7]
```diff
@@ -26,6 +36,93 @@ class QcFlagRepository extends Repository {
...
+                ON qcf.id = qcfep.flag_id
+            INNER JOIN data_pass_quality_control_flag AS dpqcf ON dpqcf.quality_control_flag_id = qcf.id
+            INNER JOIN (${gaqPeriodsSubQuery}) AS gaq_periods ON gaq_periods.data_pass_id = dpqcf.data_pass_id
+            INNER JOIN global_aggregated_quality_detectors AS gaqd
+                ON gaqd.data_pass_id = gaq_periods.data_pass_id
+                    AND gaqd.run_number = gaq_periods.run_number
+                    AND gaqd.dpl_detector_id = qcf.dpl_detector_id
+                    AND gaq_periods.run_number = qcf.run_number
+                    AND COALESCE(UNIX_TIMESTAMP(qcfep.\`from\`), 0) <= gaq_periods.\`from\`
+                        AND gaq_periods.\`to\` <= UNIX_TIMESTAMP(COALESCE(qcfep.\`to\`, NOW()))
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/database/repositories/QcFlagRepository.js]: applied

File: lib/database/repositories/QcFlagRepository.js [hunk 8]
```diff
@@ -26,6 +36,93 @@ class QcFlagRepository extends Repository {
...
+                from,
+                to,
+                contributingFlagIds,
+            }) => ({
+                dataPassId,
+                runNumber,
+                from,
+                to,
+                contributingFlagIds: contributingFlagIds.split(',').map((id) => parseInt(id, 10)),
+            }));
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/database/repositories/QcFlagRepository.js:123]: applied

File: lib/server/controllers/qcFlag.controller.js [hunk 9]
```diff
@@ -248,6 +248,34 @@ const getQcFlagsSummaryHandler = async (req, res) => {
     }
 };
 
+// eslint-disable-next-line valid-jsdoc
+/**
+ *
+ */
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/server/controllers/qcFlag.controller.js]: fixed

File: lib/server/controllers/qcFlag.controller.js (hunk 1 above)
Comment (xsalonx) [reply lib/server/controllers/qcFlag.controller.js]: fixed

File: lib/server/services/qualityControlFlag/QcFlagService.js (hunk 2 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagService.js]: artifact, revoked

File: lib/database/repositories/QcFlagRepository.js (hunk 4 above)
Comment (xsalonx) [reply lib/database/repositories/QcFlagRepository.js:46]: renamed

File: lib/server/services/qualityControlFlag/QcFlagService.js (hunk 5 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagService.js:570]: because, simply, one request can be done instead of many

File: lib/server/services/qualityControlFlag/QcFlagService.js (hunk 3 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagService.js]: because it, effective periods (but different than QC flag effective period) in which there is list of QC flags contributing to GAQ, but I renamed it

Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagService.js]: changed to `GaqPeriods`

File: lib/database/repositories/QcFlagRepository.js [hunk 10]
```diff
@@ -26,6 +36,93 @@ class QcFlagRepository extends Repository {
         super(QcFlag);
     }
 
+    /**
+     * Find GAQ effective periods for given data pass and run
+     *
+     * @param {number} dataPassId id of data pass id
+     * @param {number[]} runNumber run number
+     * @return {Promise<GaqPeriod>} promise resolvesd once periods are calculated
```
Comment (martinboulais) [inline lib/database/repositories/QcFlagRepository.js:44]:
```suggestion
     * @return {Promise<GaqPeriod[]>} Resolves with the GAQ periods
```

File: lib/database/repositories/QcFlagRepository.js [hunk 11]
```diff
@@ -26,6 +36,93 @@ class QcFlagRepository extends Repository {
         super(QcFlag);
     }
 
+    /**
+     * Find GAQ effective periods for given data pass and run
+     *
+     * @param {number} dataPassId id of data pass id
+     * @param {number[]} runNumber run number
```
Comment (martinboulais) [inline lib/database/repositories/QcFlagRepository.js:43]:
```suggestion
     * @param {number} runNumber run number
```

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/server/services/qualityControlFlag/QcFlagService.js [hunk 12]
```diff
@@ -85,6 +85,14 @@ const QC_SUMMARY_PROPERTIES = {
  * @type {Object<number, Object<number, RunQcSummary>>} runNumber to RunQcSummary mapping
  */
 
+/**
+ * @typedef GaqFlags
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagService.js:89]: this would be Global aggregated quality Quality control flags, think Gaq is enough
